pkg/protocol/engine/mempool/spenddag/spenddagv1/spend_test.go

```diff
@@ -0,0 +1,622 @@
+
+	for _, spend := range spends {
+		if !unPreferredConflicts.Has(spend) {
+			preferredConflicts.Add(spend)
+			spend.ConflictingSpends.Range(func(conflictingSpend *Spend[iotago.TransactionID, iotago.OutputID, vote.MockedRank]) {
+				if spend != conflictingSpend {
+					unPreferredConflicts.Add(conflictingSpend)
+				}
+			}, true)
+		}
+	}
+
+	for _, Spend := range spends {
+		if preferredConflicts.Has(Spend) {
+			require.True(t, Spend.IsPreferred(), "Spend %s should be preferred", Spend.ID)
+		}
+		if unPreferredConflicts.Has(Spend) {
+			require.False(t, Spend.IsPreferred(), "Spend %s should be unPreferred", Spend.ID)
+		}
+	}
+
+	_ = unPreferredConflicts.ForEach(func(unPreferredConflict TestSpend) (err error) {
+		// iterating in descending order, so the first preferred Spend
+		return unPreferredConflict.ConflictingSpends.ForEach(func(SpendingConflict TestSpend) error {
+			if SpendingConflict != unPreferredConflict && SpendingConflict.IsPreferred() {
+				require.Equal(t, SpendingConflict, unPreferredConflict.PreferredInstead())
+
+				return ierrors.New("break the loop")
+			}
+
+			return nil
+		}, true)
+	})
+}
+
+func generateRandomConflictPermutation() func(Spend TestSpend) {
+	updateType := rand.Intn(100)
+	delta := rand.Intn(100)
+
+	return func(Spend TestSpend) {
+		if updateType%2 == 0 {
+			Spend.Weight.AddCumulativeWeight(int64(delta))
+		} else {
+			Spend.Weight.RemoveCumulativeWeight(int64(delta))
+		}
+	}
+}
+
+func createConflicts(pendingTasks *syncutils.Counter) map[string]TestSpend {
+	weights := account.NewSeatedAccounts(account.NewAccounts())
+
+	thresholdProvider := acceptance.ThresholdProvider(func() int64 {
+		return int64(weights.SeatCount())
+	})
+
+	red := NewTestConflictSet(id("red"))
+	blue := NewTestConflictSet(id("blue"))
+	green := NewTestConflictSet(id("green"))
+	yellow := NewTestConflictSet(id("yellow"))
+
+	SpendA := NewTestSpend(transactionID("A"), nil, ds.NewSet(red), weight.New(), pendingTasks, thresholdProvider)
+	SpendB := NewTestSpend(transactionID("B"), nil, ds.NewSet(red, blue), weight.New(), pendingTasks, thresholdProvider)
+	SpendC := NewTestSpend(transactionID("C"), nil, ds.NewSet(green, blue), weight.New(), pendingTasks, thresholdProvider)
+	SpendD := NewTestSpend(transactionID("D"), nil, ds.NewSet(green, yellow), weight.New(), pendingTasks, thresholdProvider)
+	SpendE := NewTestSpend(transactionID("E"), nil, ds.NewSet(yellow), weight.New(), pendingTasks, thresholdProvider)
+
+	return map[string]TestSpend{
+		"SpendA": SpendA,
+		"SpendB": SpendB,
+		"SpendC": SpendC,
+		"SpendD": SpendD,
+		"SpendE": SpendE,
+	}
+}
+
+func assertPreferredInstead(t *testing.T, preferredInsteadMap map[TestSpend]TestSpend) {
+	for Spend, preferredInsteadConflict := range preferredInsteadMap {
+		assert.Equalf(t, preferredInsteadConflict.ID, Spend.PreferredInstead().ID, "Spend %s should prefer %s instead of %s", Spend.ID, preferredInsteadConflict.ID, Spend.PreferredInstead().ID)
```

Comment:
```suggestion
	for Spend, preferredInsteadSpend := range preferredInsteadMap {
		assert.Equalf(t, preferredInsteadSpend.ID, Spend.PreferredInstead().ID, "Spend %s should prefer %s instead of %s", Spend.ID, preferredInsteadConflict.ID, Spend.PreferredInstead().ID)
```